src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -208,30 +232,62 @@ CoreWorkerDirectTaskSubmitter::GetOrConnectLeaseClient(
 
 void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
     const SchedulingKey &scheduling_key, const rpc::Address *raylet_address) {
-  if (pending_lease_requests_.find(scheduling_key) != pending_lease_requests_.end()) {
+  auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+  auto &pending_lease_request = scheduling_key_entry.pending_lease_request_;
+
+  if (pending_lease_request.first) {
     // There's already an outstanding lease request for this type of task.
     return;
   }
-  auto it = task_queues_.find(scheduling_key);
-  if (it == task_queues_.end()) {
+
+  auto &task_queue = scheduling_key_entry.task_queue_;
+  if (task_queue.empty()) {
     // We don't have any of this type of task to run.
     return;
   }
 
+  // If pipelining is enabled, check whether we really need a new worer or whether we have
+  // enough room in an existing worker's pipeline to send the new tasks
+  if (max_tasks_in_flight_per_worker_ > 1) {
+    if (scheduling_key_entry.tot_tasks_in_flight <
+        scheduling_key_entry.active_workers_.size() * max_tasks_in_flight_per_worker_) {
```

Comment:
I'm not sure about putting this logic in `RequestNewWorkerIfNeeded`. It seems a bit strange to be submitting tasks in this method (before the logic only had to do with requesting new workers and not anything to do with task submission). It also seems a bit brittle because `OnWorkerIdle`, called below, calls back into `RequestNewWorkerIfNeeded`.

Instead, how about we only use the check to see whether we should request a new worker or not? Then, we should also move this new logic to submit tasks to already active workers to when the task is first queued in `SubmitTask`.

Reply:
The reason why I added this part (calling OnWorkerIdle from within RequestNewWorkerIfNeeded) was to avoid introducing new latency due to the fact that `OnWorkerIdle` works in a _pull_ (rather than _push_) fashion. `OnWorkerIdle` is normally only called when (1) we get a new worker, or (2) when we get a response from a worker (after the worker has completed the execution of a task). At that point, we pull tasks from the queue and submit them.  
Now, consider a scenario with a set of active workers with non-full pipelines, where we have just added a few more tasks to a queue at the owner's. In this situation, even if some of the pipelines are not full, before we can submit the new tasks to an active worker, we have to wait until that worker has responded back to the owner (because OnWorkerIdle is not called until then). If we submit to a queue at the owner's a number of tasks that is larger than the grand total of existing "spots" available in the pipelines to the existing worker, the system will only fill those pipelines (and realize that it needs to request an additional worker) only after it has received a response from every single one of the existing workers with non-full pipelines. As a result, the total number of active workers will be lower for a longer period of time, and the overall execution time will suffer.

Reply:
Yes, I understand that we should call `OnWorkerIdle` even while it is still executing other tasks. My comment was more about where we should call it. I think it is better to call it directly in `SubmitTask` right after we've added tasks to the queue. The reason is that: a) dispatching tasks doesn't match the current semantics of the method, which is supposed to only request a new worker, and b) it prevents the bad recursive structure where `RequestNewWorkerIfNeeded` calls `OnWorkerIdle` calls `RequestNewWorkerIfNeeded`, etc.

Reply:
Oh I see! I think I misread your initial comment! Sorry about that. This makes sense. So I guess we would just move the code block 

```cpp
if (scheduling_key_entry.tot_tasks_in_flight <
      scheduling_key_entry.active_workers_.size() * max_tasks_in_flight_per_worker_) {
    // The pipelines to the current workers are not full yet, so we don't need more
    // workers.

    // Find a worker with a number of tasks in flight that is less than the maximum
    // value (max_tasks_in_flight_per_worker_) and call OnWorkerIdle to send tasks to
    // that worker
    for (auto active_worker_addr : scheduling_key_entry.active_workers_) {
      RAY_CHECK(worker_to_lease_entry_.find(active_worker_addr) !=
                worker_to_lease_entry_.end());
      auto &lease_entry = worker_to_lease_entry_[active_worker_addr];
      if (lease_entry.tasks_in_flight_ < max_tasks_in_flight_per_worker_) {
        OnWorkerIdle(active_worker_addr, scheduling_key, false,
                     lease_entry.assigned_resources_);
        break;
      }
    }

    return;
  } 
```

(without the return statement, of course) to the `SubmitTask` function. `RequestNewWorkerIfNeeded` would then only keep the following if statement?

```cpp
if (scheduling_key_entry.tot_tasks_in_flight <
      scheduling_key_entry.active_workers_.size() * max_tasks_in_flight_per_worker_) {
    // The pipelines to the current workers are not full yet, so we don't need more
    // workers.

    return;
  } 
```

Reply:
Yes, exactly!

Reply:
Sounds good! Let me do that right now.